nautobot/dcim/forms.py

```diff
@@ -5157,3 +5176,117 @@ class Meta:
             "weight",
             "tags",
         )
+
+
+#
+# Virtual Device Context
+#
+
+
+class VirtualDeviceContextForm(NautobotModelForm):
+    device = DynamicModelChoiceField(
+        queryset=Device.objects.all(),
+    )
+    tenant = DynamicModelChoiceField(
+        queryset=Tenant.objects.all(),
+        required=False,
+    )
+    interfaces = DynamicModelMultipleChoiceField(
+        queryset=Interface.objects.all(), required=False, query_params={"device": "$device"}
+    )
+    primary_ip4 = DynamicModelChoiceField(
+        queryset=IPAddress.objects.all(), required=False, query_params={"ip_version": 4, "interfaces": "$interfaces"}
+    )
+    primary_ip6 = DynamicModelChoiceField(
+        queryset=IPAddress.objects.all(), required=False, query_params={"ip_version": 6, "interfaces": "$interfaces"}
+    )
+    status = DynamicModelChoiceField(
+        queryset=Status.objects.all(),
+        required=True,
+        query_params={"content_types": VirtualDeviceContext._meta.label_lower},
+    )
+
+    class Meta:
+        model = VirtualDeviceContext
+        fields = [
+            "name",
+            "device",
+            "status",
+            "identifier",
+            "interfaces",
+            "primary_ip4",
+            "primary_ip6",
+            "tenant",
+            "description",
+            "tags",
+        ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.instance.present_in_database:
+            # We need to set the initial value for 'interfaces' because, by default,
+            # ModelForms only handle fields directly defined on the model (e.g., the 'name' field).
+            self.fields["interfaces"].initial = self.instance.interfaces.all()
+        else:
+            # An object that doesn't exist yet can't have any Primary IPs
+            self.fields["primary_ip4"].widget.attrs["disabled"] = True
+            self.fields["primary_ip6"].widget.attrs["disabled"] = True
+
+    def save(self, commit=True):
+        data = super().save(commit)
+        interfaces = self.cleaned_data["interfaces"]
+        self.instance.interfaces.set(interfaces)
```

Comment:
Would we need to defer setting/validating the `self.instance.primary_ip[46]` values until after we've updated/validated the `self.instance.interfaces`? Feels like there's some sort of chicken-and-egg problem here, because otherwise couldn't we fail to set the primary_ip from an interface not yet assigned, or permit setting the primary_ip for an interface that's about to be removed?